From 6ddc2e5eb5565c8841c2ef047642fa766e1a366a Mon Sep 17 00:00:00 2001
From: Lawrence Rust <lvr@softsystem.co.uk>
Date: Mon, 7 Jul 2014 18:53:47 +0100
Subject: [PATCH 11/56] MythMusic: Avoid a floating point domain error in the
 spectrum visualiser

Avoid taking the log of 0

This change avoids the 'sticky bars' sometimes seen with the spectrum
and squares visualisers.

Signed-off-by: Paul Harrison <pharrison@mythtv.org>
---
 mythplugins/mythmusic/mythmusic/visualize.cpp |    9 +++++----
 1 file changed, 5 insertions(+), 4 deletions(-)

diff --git a/mythplugins/mythmusic/mythmusic/visualize.cpp b/mythplugins/mythmusic/mythmusic/visualize.cpp
index d9b4c90..79be5a1 100644
--- a/mythplugins/mythmusic/mythmusic/visualize.cpp
+++ b/mythplugins/mythmusic/mythmusic/visualize.cpp
@@ -697,10 +697,11 @@ bool Spectrum::process(VisualNode *node)
 
     for (i = 0; (int)i < rects.size(); i++, w += analyzerBarWidth)
     {
-        magL = (log(sq(real(lout[index])) + sq(real(lout[FFTW_N - index]))) - 22.0) *
-               scaleFactor;
-        magR = (log(sq(real(rout[index])) + sq(real(rout[FFTW_N - index]))) - 22.0) *
-               scaleFactor;
+        tmp = sq(real(lout[index])) + sq(real(lout[FFTW_N - index]));
+        magL = (tmp > 1.) ? (log(tmp) - 22.0) * scaleFactor : 0.;
+
+        tmp = sq(real(rout[index])) + sq(real(rout[FFTW_N - index]));
+        magR = (tmp > 1.) ? (log(tmp) - 22.0) * scaleFactor : 0.;
 
         if (magL > size.height() / 2)
         {
-- 
1.7.10.2

